### PR TITLE
feat(activity-logs): track authenticated page visits (USER_VISITED)

### LINKED ISSUE #110: feat(activity-logs): track authenticated page visits (USER_VISITED) — clean restart
## Objective

Add a `USER_VISITED` activity log entry that fires **once per UTC day per user** when they visit any `/dashboard/*` route. This closes a gap where users with active sessions (logged in yesterday, visiting today) leave no trace in the activity log — `USER_SIGNED_IN` only fires on session creation.

**Keep it simple.** This is a small feature (~5 files, ~35 lines). Do not expand scope.

---

## ⚠️ Critical context — READ BEFORE TOUCHING ANYTHING

The previous attempt (#108, closed) got tangled up in Prisma migration infrastructure work. That infrastructure is **already deployed** — do not redo it.

### Database state (both local + production)

The `_prisma_migrations` table on both DBs contains these records, both marked as **applied**:

| Migration | Status |
|---|---|
| `0001_init` | applied |
| `0002_add_missing_schema` | applied |

### Code state

- `package.json` build script uses `prisma migrate deploy` (not `db push`)
- Two migration files are needed in `prisma/migrations/` to stay consistent with the DB records
- **Both the migration files and the build script change exist on branch `chore/db-migration-infrastructure`**

### ❌ DO NOT

- Delete or modify the migration files — the DB records reference them
- Run `prisma db push` — the workflow is now migration-based
- Run `prisma migrate resolve` against production unless absolutely required
- Rewrite `package.json` build script back to `db push`
- Expand scope beyond the `USER_VISITED` feature

### ✅ DO

- Merge branch `chore/db-migration-infrastructure` to `main` **first**. This PR should be reviewed and merged before starting the feature work. It contains only:
  - `prisma/migrations/0001_init/migration.sql`
  - `prisma/migrations/0002_add_missing_schema/migration.sql`
  - `package.json` (build script change)
- After that's merged, implement the `USER_VISITED` feature on a fresh branch off `main`

---

### Step 1 — Add enum value

In `prisma/schema.prisma`, add `USER_VISITED` to the `ActivityAction` enum:

```prisma
enum ActivityAction {
  USER_SIGNED_IN
  USER_SIGNED_OUT
  SIGN_IN_FAILED
  OAUTH_ACCOUNT_LINKED
  PASSWORD_CHANGED
  PASSWORD_RESET
  USER_VISITED
}
```

Then run: `bunx prisma migrate dev --name add_user_visited_action` and commit the generated migration file.

### Step 2 — Track visits in `proxy.ts`

Modify `proxy.ts` to log a `USER_VISITED` event on authenticated `/dashboard/*` visits, throttled once per UTC day via an httpOnly cookie.

**Key points:**
- Use `logActivity()` from `@/lib/activity-log` — fire-and-forget (`void logActivity(...)`)
- Cookie name: `vl`, value: today's UTC date (`YYYY-MM-DD`), `httpOnly`, `sameSite: "lax"`, `maxAge: 86400`
- Cookie is only set on responses for authenticated protected routes
- Never block the request path

Full diff for `proxy.ts`:

```diff
 import { getCookies } from "better-auth/cookies";
 import type { NextRequest } from "next/server";
 import { NextResponse } from "next/server";
+import { extractRequestMeta, logActivity } from "@/lib/activity-log";
 import { auth } from "@/lib/auth";
 import { safeCallbackUrl, sanitizeCallbackUrl } from "@/lib/auth/safe-callback";
 import { prisma } from "@/lib/db";

 const { sessionToken } = getCookies(auth.options);

+// Cookie set once per UTC day to throttle USER_VISITED log writes
+const VISIT_COOKIE = "vl";
+
 async function resolveSession(request: NextRequest) {
 	try {
 		return await auth.api.getSession({ headers: request.headers });
@@ -29,6 +33,9 @@ export async function proxy(request: NextRequest) {
 		return NextResponse.redirect(loginUrl);
 	}

+	const todayUtc = new Date().toISOString().slice(0, 10);
+	let pendingVisitActorId: string | null = null;
+
 	if (isProtected && sessionCookie) {
 		const session = await resolveSession(request);
 		if (!session) {
@@ -45,6 +52,10 @@ export async function proxy(request: NextRequest) {
 			response.cookies.delete(sessionToken.name);
 			return response;
 		}
+
+		if (request.cookies.get(VISIT_COOKIE)?.value !== todayUtc) {
+			pendingVisitActorId = session.user.id;
+		}
 	}

 	if ((pathname === "/login" || pathname === "/register") && sessionCookie) {
@@ -63,7 +74,25 @@ export async function proxy(request: NextRequest) {
 		return NextResponse.redirect(new URL(safeCallback, request.url));
 	}

-	return NextResponse.next();
+	const response = NextResponse.next();
+
+	if (pendingVisitActorId) {
+		const { ipAddress, userAgent } = extractRequestMeta(request.headers);
+		void logActivity({
+			action: "USER_VISITED",
+			actorId: pendingVisitActorId,
+			ipAddress,
+			userAgent,
+		});
+		response.cookies.set(VISIT_COOKIE, todayUtc, {
+			path: "/",
+			httpOnly: true,
+			sameSite: "lax",
+			maxAge: 86400,
+		});
+	}

+	return response;
 }
```

### Step 3 — Update admin activity logs UI

Three files. Each needs `USER_VISITED: "Visited"` added to the existing `ACTION_LABELS` object, and `"USER_VISITED"` added to the `ACTIONS` array.

**`app/(dashboard)/dashboard/admin-settings/activity-logs/_components/activity-log-table.tsx`:**
```diff
 const ACTION_LABELS: Record<ActivityAction, string> = {
 	USER_SIGNED_IN: "Signed in",
 	USER_SIGNED_OUT: "Signed out",
 	SIGN_IN_FAILED: "Sign-in failed",
 	OAUTH_ACCOUNT_LINKED: "OAuth linked",
 	PASSWORD_CHANGED: "Password changed",
 	PASSWORD_RESET: "Password reset",
+	USER_VISITED: "Visited",
 };
```

**`app/(dashboard)/dashboard/admin-settings/activity-logs/_components/activity-log-filters.tsx`:**
Same change as above.

**`app/(dashboard)/dashboard/admin-settings/activity-logs/page.tsx`:**
```diff
 const ACTIONS: ActivityAction[] = [
 	"USER_SIGNED_IN",
 	"USER_SIGNED_OUT",
 	"SIGN_IN_FAILED",
 	"OAUTH_ACCOUNT_LINKED",
 	"PASSWORD_CHANGED",
 	"PASSWORD_RESET",
+	"USER_VISITED",
 ];
```

---

## Required workflow

1. Review + merge `chore/db-migration-infrastructure` → `main` (separate PR, infra only)
2. Create `feature/user-visited-tracking` off `main`
3. Make the 5 file changes above + generate a tiny migration for the enum addition
4. **Run `bun run lint` before committing** (CI checks Biome — was missed last time)
5. Open a single-purpose PR

## Test plan

- [ ] Sign in, visit multiple dashboard pages → only one `USER_VISITED` entry for the day
- [ ] Log in today, close browser, reopen tomorrow → new `USER_VISITED` entry
- [ ] User with valid session from yesterday visits `/dashboard` → `USER_VISITED` logged
- [ ] `USER_VISITED` appears in the admin activity logs filter dropdown with label "Visited"
- [ ] Existing auth events (`USER_SIGNED_IN`, `USER_SIGNED_OUT`, etc.) unaffected
- [ ] Unauthenticated visits to `/dashboard` redirect to login — no log entry
- [ ] `bun run lint` passes
- [ ] `bun run build` passes locally
- [ ] Vercel preview deploy passes

## Related (closed)

- Original issue: #108 (closed)
- Original PR: #109 (closed) — got tangled up in migration infrastructure work that's now already done

## Note to the next agent

This feature is small. Write the 5 file changes, create the migration for the enum, run lint, commit, push, done. If you find yourself doing anything beyond that, stop and ask the user.

### PR DESCRIPTION
**Pull request by chrisgen19**
## Summary
- Adds `USER_VISITED` to the `ActivityAction` enum in `prisma/schema.prisma` and wires it through the admin activity-logs UI (filter dropdown + row label: "Visited").
- Logs a `USER_VISITED` activity entry on authenticated `/dashboard/*` visits, throttled to **once per UTC day per user** via an httpOnly `vl` cookie (`YYYY-MM-DD`, `sameSite: lax`, `maxAge: 86400`).
- Fire-and-forget (`void logActivity(...)`) — never blocks the request path.

Closes the gap where users with an active session from a prior day leave no trace in the activity log (`USER_SIGNED_IN` only fires on session creation).

## Notes

**No migration file in this PR.** The `USER_VISITED` enum value already exists in both local and production databases — it was declared inside `0002_add_missing_schema/migration.sql` (line 2) as part of the infra PR #111 that was just merged. Running `bunx prisma migrate dev` against the current schema produces no diff. A separate follow-up issue tracks a pre-existing bug in `0001_init/migration.sql` that prevents `migrate dev` from running at all; it is unrelated to this feature.

## Test plan
- [ ] Sign in, visit multiple `/dashboard/*` pages in the same day → only one `USER_VISITED` row per user
- [ ] Clear the `vl` cookie → next dashboard visit produces a new `USER_VISITED` row
- [ ] User with a valid session from a prior day visits `/dashboard` today → `USER_VISITED` logged
- [ ] `USER_VISITED` appears in the admin activity-logs filter dropdown with label "Visited"
- [ ] Existing auth events (`USER_SIGNED_IN`, `USER_SIGNED_OUT`, etc.) unaffected
- [ ] Unauthenticated visit to `/dashboard` redirects to login and logs nothing
- [x] `bun run lint` passes locally
- [ ] Vercel preview build passes
- [ ] CI (unit + E2E) passes

Closes #110

### CODE REVIEW

**Review by chrisgen19 — COMMENTED** [review]
Code review findings (actionable):

1) High — "Once per UTC day per user" is not actually enforced server-side
- Affected file: `proxy.ts`
- Why it matters: Dedup currently depends on the `vl` client cookie only. The same user can still generate multiple `USER_VISITED` rows in one UTC day from multiple devices/browsers or near-simultaneous first requests before cookie round-trip completion.
- Smallest safe fix: Enforce idempotency in the database for `USER_VISITED` (e.g. unique key on actor + utc_day + action, with upsert/ON CONFLICT DO NOTHING). Keep cookie throttling as an optimization.

2) Medium — Missing automated tests for visit-tracking behavior
- Affected area: `proxy.ts` behavior coverage
- Why it matters: Edge cases (UTC rollover, repeat requests, unauthenticated redirects) are easy to regress.
- Smallest safe fix: Add focused tests that mock session/db/logging for:
  - unauthenticated `/dashboard/*` redirects and logs nothing
  - first authenticated daily visit logs + sets cookie
  - same-day repeat request does not log
  - next UTC day logs again

3) Low — Visit cookie missing explicit `secure` flag
- Affected file: `proxy.ts`
- Why it matters: Without `secure`, cookie can be sent in non-TLS contexts (dev/misconfigured environments), weakening signal integrity.
- Smallest safe fix: set `secure: process.env.NODE_ENV === "production"` (or equivalent HTTPS check).

Assumption:
- I reviewed this as "once/day globally per user". If intended behavior is "once/day per browser," finding #1 severity can be lowered.

**Comment by chrisgen19** [comment]
Thanks for the review — addressed all three findings in 251209f.

**#1 (High) — Enforced server-side.** Added a partial unique index `activity_logs_user_visited_daily_key` on `(actor_id, created_at::date) WHERE action = 'USER_VISITED'`, so "once per UTC day per user" is now enforced at the DB level across devices and any cookie-throttle race. `lib/activity-log.ts` swallows `P2002` for `USER_VISITED` so the duplicate write is silent. The cookie stays as an optimization to avoid a DB round-trip on every page view.

Implementation note: the index is declared via raw SQL in `prisma/migrations/0003_user_visited_daily_unique/migration.sql`. Prisma's schema can't express partial + expression unique indexes, so this is the standard workaround. The `activity_logs` table is empty in all known environments today, so the index creates cleanly on deploy. Prisma stores timestamps as UTC by default, so `created_at::date` gives the UTC date without needing `AT TIME ZONE 'UTC'` (which Postgres rejects as non-IMMUTABLE).

**#2 (Medium) — Added.** New `proxy.test.ts` covers the four scenarios you listed:

- unauthenticated `/dashboard` redirects and logs nothing
- first authenticated daily visit logs + sets `vl` cookie
- same-day repeat request does not log and does not re-set the cookie
- `vl` cookie from a prior day (UTC rollover) re-triggers the log

**#3 (Low) — Fixed.** Cookie now sets `secure: process.env.NODE_ENV === "production"` (matching the existing pattern in `lib/db/index.ts`).

Full unit suite: 161 passing. Biome clean.

**Review by chrisgen19 — COMMENTED** [review]
Follow-up review after latest fixes:

Remaining finding:

1) Medium — "UTC day" uniqueness still depends on database timezone
- Affected files: `prisma/migrations/0003_user_visited_daily_unique/migration.sql`, `proxy.ts`
- Why it matters: uniqueness is enforced with `(created_at::date)`, but `created_at` is `TIMESTAMP(3)` (without timezone). That means day-boundary behavior can drift if DB/session timezone is not UTC, which can violate the stated "once per UTC day per user" requirement.
- Smallest safe fix: persist an explicit UTC day key from app code (e.g. `visit_day_utc DATE` derived from `new Date().toISOString().slice(0,10)`) and enforce uniqueness on `(actor_id, visit_day_utc)` for `USER_VISITED`.

What is fixed from previous review:
- Server-side dedup enforcement: fixed (partial unique index + P2002 idempotency handling)
- Missing automated tests: fixed (`proxy.test.ts` with core scenarios)
- Cookie `secure` flag: fixed

**Comment by chrisgen19** [comment]
Addressed in 6bcf944.

**Medium — DB-timezone drift — fixed.** Added explicit `visit_day_utc DATE` column on `activity_logs`, populated from `new Date().toISOString().slice(0,10)` in `logActivity()` so the value is always a JS/UTC date regardless of Postgres session TZ or `CURRENT_TIMESTAMP` behavior. The unique index is rebuilt on `(actor_id, visit_day_utc) WHERE action = 'USER_VISITED'` — plain columns, no expression, deterministic.

Changes:
- `prisma/schema.prisma` — `ActivityLog.visitDayUtc DateTime? @map("visit_day_utc") @db.Date`
- `prisma/migrations/0004_activity_log_visit_day_utc/migration.sql` — adds the column, drops the old expression index, creates the new column-based index
- `lib/activity-log.ts` — computes and writes `visitDayUtc` for `USER_VISITED` action; `null` for everything else
- Migration applied locally, regenerated Prisma client, 161/161 tests still passing, Biome clean

The column is nullable so existing rows (none for `USER_VISITED` yet, but future non-`USER_VISITED` rows) don't need a backfill and the default `null` coexists with the partial unique index without issue.

**Comment by chrisgen19** [comment]
Addressed M1 and M3 from the self-review in 7c76d96.

**M1 — Duplicate Prisma import.** Consolidated to a single import: `import { type ActivityAction, Prisma } from "@/app/generated/prisma/client"`. `Prisma.PrismaClientKnownRequestError` is now used directly; `Prisma.InputJsonValue` still works as a type under the value import.

**M3 — Test coverage for P2002 swallow.** New `lib/activity-log.test.ts` (5 tests) locks in the idempotency contract:
- `USER_VISITED` writes populate `visit_day_utc` with UTC midnight
- Non-`USER_VISITED` writes leave `visit_day_utc` null
- P2002 on `USER_VISITED` is silently swallowed (no `console.error`)
- P2002 on other actions is still logged as an error
- Non-P2002 errors on `USER_VISITED` are logged, not swallowed

Skipped M2 (write-path TZ defensive rewrite) — the production stack (Vercel Node 20 + Postgres with UTC session) is known-UTC and the explicit-column fix already eliminates the read-side ambiguity. Happy to revisit if it comes up.

166/166 unit tests passing. Biome clean.
